### PR TITLE
[autopilot] penalize small channels in preferantial attachment heuristic

### PR DESCRIPTION
**Pull request by halseth**
To avoid assigning a high score to nodes with a large number of small
channels, we only count channels at least as large as a given fraction of
the graph's median channel size.

### CODE REVIEW

File: autopilot/prefattach.go [hunk 1]
```diff
@@ -72,12 +77,63 @@ func (p *PrefAttachment) NodeScores(g ChannelGraph, chans []Channel,
...
+	// the graph.
 	var maxChans int
 	nodeChanNum := make(map[NodeID]int)
 	if err := g.ForEachNode(func(n Node) error {
 		var nodeChans int
-		err := n.ForEachChannel(func(_ ChannelEdge) error {
+		err := n.ForEachChannel(func(e ChannelEdge) error {
+			// Since connecting to nodes with a lot of small
+			// channels actually worsens our connectivity in the
+			// graph (we will potentially waste time tryng to use
```
Comment (alexbosworth) [inline autopilot/prefattach.go]:
```suggestion
			// graph (we will potentially waste time trying to use
```

File: rpcserver.go [hunk 2]
```diff
@@ -3961,6 +3967,20 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
 		return nil, err
 	}
 
+	// Sort the channels by capacity, and find the median.
```
Comment (alexbosworth) [inline rpcserver.go]: Maybe there should be a cap here in case the median is super high for some reason?

Comment (halseth) [reply rpcserver.go]: You mean cap the number of channels we'll sort?

Comment (alexbosworth) [reply rpcserver.go]: Cap the median value amount

Comment (Roasbeef) [reply rpcserver.go]: This is just for the RPC call, why should we cap it?

File: rpcserver.go [hunk 3]
```diff
@@ -3961,6 +3967,20 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
...
+	sort.Slice(allChans, func(i, j int) bool {
+		return allChans[i] < allChans[j]
+	})
+
+	if numChannels == 0 {
```
Comment (Roasbeef) [inline rpcserver.go]: Switch statement instead perhaps?

File: autopilot/prefattach.go [hunk 4]
```diff
@@ -72,12 +77,63 @@ func (p *PrefAttachment) NodeScores(g ChannelGraph, chans []Channel,
...
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	sort.Slice(allChans, func(i, j int) bool {
```
Comment (Roasbeef) [inline autopilot/prefattach.go]: Duplicate code to what was added in the server, we can add an abstract method in the `router` package perhaps? Or a "static method" (lol) on the `ChannelGraph`

File: autopilot/prefattach.go [hunk 5]
```diff
@@ -72,12 +77,63 @@ func (p *PrefAttachment) NodeScores(g ChannelGraph, chans []Channel,
...
 	if err := g.ForEachNode(func(n Node) error {
 		var nodeChans int
-		err := n.ForEachChannel(func(_ ChannelEdge) error {
+		err := n.ForEachChannel(func(e ChannelEdge) error {
+			// Since connecting to nodes with a lot of small
+			// channels actually worsens our connectivity in the
+			// graph (we will potentially waste time tryng to use
+			// these useless channels in path finding), we decrease
+			// the counter for such channels.
+			if e.Capacity < medianChanSize/4 {
```
Comment (Roasbeef) [inline autopilot/prefattach.go]: `4` should be moved up into a constant

File: autopilot/prefattach.go (hunk 1 above)
Comment (halseth) [reply autopilot/prefattach.go]: Fixed.

File: rpcserver.go (hunk 2 above)
Comment (halseth) [reply rpcserver.go]: Question is whether to cap it for prefattach?

File: rpcserver.go (hunk 3 above)
Comment (halseth) [reply rpcserver.go]: Fixed.

File: autopilot/prefattach.go (hunk 4 above)
Comment (halseth) [reply autopilot/prefattach.go]: added `autopilot.Median()`

File: autopilot/prefattach.go (hunk 5 above)
Comment (halseth) [reply autopilot/prefattach.go]: Done.

**Comment by halseth** [comment]
PTAL @alexbosworth @Roasbeef

File: autopilot/prefattach.go [hunk 6]
```diff
@@ -72,12 +82,50 @@ func (p *PrefAttachment) NodeScores(g ChannelGraph, chans []Channel,
...
 
-	// Count the number of channels for each particular node in the graph.
+	// We first run though the graph once in order to find the median
+	// channel size.
+	var (
+		allChans  []btcutil.Amount
+		seenChans = make(map[uint64]struct{})
+	)
+	if err := g.ForEachNode(func(n Node) error {
+		err := n.ForEachChannel(func(e ChannelEdge) error {
```
Comment (alexbosworth) [inline autopilot/prefattach.go]: It may be safer to only consider channels that have recently been updated

**Comment by alexbosworth** [comment]
lgtm although I still feel iffy about situations where the median could actually yield unexpected results. The code in comments says "large" but median doesn't necessarily imply large

**Comment by halseth** [comment]
@alexbosworth "large-ish channels" 😛

**Pull request merged**